src/commonMain/kotlin/org/kbson/serialization/Serializers.kt

```diff
@@ -0,0 +1,799 @@
+                }
+                jsonElement.intOrNull?.let {
+                    return BsonInt32(it)
+                }
+                jsonElement.floatOrNull?.let {
+                    return BsonDouble(it.toDouble())
+                }
+                jsonElement.doubleOrNull?.let {
+                    return BsonDouble(it)
+                }
+                jsonElement.contentOrNull?.let {
+                    return BsonString(it)
+                }
+                return BsonNull
+            }
+            is JsonNull -> return BsonNull
+            else -> throw SerializationException("Unknown jsonElement type: $jsonElement")
+        }
+    }
+
+    @Suppress("ComplexMethod")
+    private fun deserializationStrategy(element: JsonObject): DeserializationStrategy<out BsonValue>? {
+        if (element.keys.isEmpty()) return null
+        return when (element.keys.first()) {
+            "\$binary" -> BsonBinarySerializer
+            "\$code" -> if ("\$scope" in element) BsonJavaScriptWithScopeSerializer else BsonJavaScriptSerializer
+            "\$date" -> BsonDateTimeSerializer
+            "\$dbPointer" -> BsonDBPointerSerializer
+            "\$maxKey" -> BsonMaxKeySerializer
+            "\$minKey" -> BsonMinKeySerializer
+            "\$numberDecimal" -> BsonDecimal128Serializer
+            "\$numberDouble" -> BsonDoubleSerializer
+            "\$numberInt" -> BsonInt32Serializer
+            "\$numberLong" -> BsonInt64Serializer
+            "\$oid" -> BsonObjectIdSerializer
+            "\$regularExpression" -> BsonRegularExpressionSerializer
+            "\$symbol" -> BsonSymbolSerializer
+            "\$timestamp" -> BsonTimestampSerializer
+            "\$undefined" -> BsonUndefinedSerializer
+            else -> null
+        }
+    }
+}
+
+@OptIn(ExperimentalSerializationApi::class)
+@Serializer(forClass = BsonDocument::class)
+private object BsonDocumentSerializer : KSerializer<BsonDocument> {
+
+    val serializer = MapSerializer(BsonDocumentKeySerializer, BsonValueSerializer)
+    override val descriptor: SerialDescriptor = SerialDescriptor("BsonDocument", serializer.descriptor)
+
+    override fun serialize(encoder: Encoder, value: BsonDocument) {
+        when (encoder) {
+            is JsonEncoder -> serializer.serialize(encoder, value)
+            else -> throw SerializationException("Unknown encoder type: $encoder")
+        }
+    }
+
+    override fun deserialize(decoder: Decoder): BsonDocument {
+        return when (decoder) {
+            is JsonDecoder -> BsonValueSerializer.deserialize(decoder).asDocument()
+            else -> throw SerializationException("Unknown decoder type: $decoder")
+        }
+    }
+}
+
+private object BsonDocumentKeySerializer : KSerializer<String> {
+    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("BsonDocumentKey", PrimitiveKind.STRING)
+
+    override fun deserialize(decoder: Decoder): String {
+        TODO("Not yet implemented")
```

Comment:
I'm not clear how we round trip keys without implementing this?  Can you explain?

Reply:
I've removed the todo - so this could be used if required.  Decoding we use JsonElements and that handles the decoding of the keys.  See `BsonValueSerializer.decoder`.